src/sentry/sentry_metrics/querying/api.py

```diff
@@ -289,44 +348,130 @@ def _build_request(self, query: MetricsQuery) -> Request:
-                )
-                totals_result = run_query(request=self._build_request(query))
+            query = executable_query.query
+
+            series_result = None
+            if executable_query.with_series:
+                series_result = run_query(request=self._build_request(query))
+
+            totals_result = None
+            if executable_query.with_totals:
+                # In case we have a series query, we want to align the query intervals so that the totals align. This
+                # is not needed when running a single totals query.
+                if series_result:
+                    modified_start = series_result["modified_start"]
+                    modified_end = series_result["modified_end"]
+                    query = (
+                        executable_query.query.set_start(modified_start)
+                        .set_end(modified_end)
+                        .set_rollup(Rollup(totals=True))
+                    )
 
-                return {**series_result, "totals": totals_result["data"]}
+                totals_result = run_query(request=self._build_request(query))
 
-            return series_result
+            result = {}
+            if series_result and totals_result:
+                result = {
+                    "series": series_result,
+                    "totals": totals_result,
+                    "modified_start": series_result["modified_start"],
+                    "modified_end": series_result["modified_end"],
+                }
+            elif series_result:
+                result = {
+                    "series": series_result,
+                    "modified_start": series_result["modified_start"],
+                    "modified_end": series_result["modified_end"],
+                }
+            elif totals_result:
+                result = {
+                    "totals": totals_result,
+                    "modified_start": totals_result["modified_start"],
+                    "modified_end": totals_result["modified_end"],
+                }
+
+            return executable_query.build_result(result)
         except SnubaError as e:
             sentry_sdk.capture_exception(e)
             raise MetricsQueryExecutionError(
                 f"An error occurred while executing the query: {type(e).__name__}"
             )
 
-    def schedule(self, query: MetricsQuery, with_totals: bool = True):
-        # By default we want to execute totals.
-        self._scheduled_queries.append((query, with_totals))
+    def schedule(self, query: MetricsQuery, group_bys: Optional[Sequence[str]]):
+        executable_query = ExecutableQuery(
+            query=query,
+            group_bys=group_bys,
+            # For now, we execute both queries independently of the query.
+            with_series=True,
+            with_totals=True,
+        )
+        self._scheduled_queries.append(executable_query)
+
+    def _derive_optimal_interval(self, result: ExecutionResult) -> int:
```

Comment:
nit: this naming is a bit confusing since inside this function optimal_interval refers to the mathematically optimal interval that is actually not returned. Maybe we could name it something like `derive_larger_interval`